cpu/stm32/stmclk/stmclk_l4wb.c

```diff
@@ -226,68 +326,90 @@ void stmclk_init_sysclk(void)
-        }
-        /* Select MSI as system clock and configure the different prescalers */
-        RCC->CFGR = (RCC_CFGR_SW_MSI | CLOCK_AHB_DIV | CLOCK_APB1_DIV | CLOCK_APB2_DIV);
     }
-    else if (IS_ACTIVE(CONFIG_USE_CLOCK_PLL)) {
-        if (IS_ACTIVE(CONFIG_CLOCK_PLL_SRC_MSI)) {
-            /* reset clock to MSI with 48MHz, disables all other clocks */
-#if defined(CPU_FAM_STM32WB)
-            RCC->CR |= (CLOCK_MSIRANGE | RCC_CR_MSION);
-#else
-            RCC->CR |= (CLOCK_MSIRANGE | RCC_CR_MSION | RCC_CR_MSIRGSEL);
-#endif
-            while (!(RCC->CR & RCC_CR_MSIRDY)) {}
 
-            if (IS_ACTIVE(CONFIG_BOARD_HAS_LSE)) {
-                /* configure the low speed clock domain */
-                stmclk_enable_lfclk();
-                /* now we can enable the MSI PLL mode to enhance accuracy of the MSI */
-                RCC->CR |= RCC_CR_MSIPLLEN;
-                while (!(RCC->CR & RCC_CR_MSIRDY)) {}
-            }
-
-            if (CONFIG_CLOCK_MSI == MHZ(48)) {
-                /* select the MSI clock for the 48MHz clock tree (USB, RNG) */
-                RCC->CCIPR = (RCC_CCIPR_CLK48SEL_0 | RCC_CCIPR_CLK48SEL_1);
-            }
+    if (IS_ACTIVE(CLOCK_ENABLE_PLL)) {
+        if (IS_ACTIVE(CONFIG_CLOCK_PLL_SRC_MSI) && IS_ACTIVE(CONFIG_BOARD_HAS_LSE)) {
+            /* configure the low speed clock domain */
+            stmclk_enable_lfclk();
+            /* now we can enable the MSI PLL mode to enhance accuracy of the MSI */
+            RCC->CR |= RCC_CR_MSIPLLEN;
+            while (!(RCC->CR & RCC_CR_MSIRDY)) {}
         }
 
         /* now we can safely configure and start the PLL */
-        RCC->PLLCFGR = (PLL_SRC | PLL_M | PLL_N | PLL_R | RCC_PLLCFGR_PLLREN);
+        RCC->PLLCFGR = (PLL_SRC | PLL_M | PLL_N | PLL_R | PLL_Q);
+        if (IS_ACTIVE(CONFIG_USE_CLOCK_PLL)) {
+            /* Enable PLLCLK if PLL is used as system clock */
+            RCC->PLLCFGR |= RCC_PLLCFGR_PLLREN;
+        }
+
+        if (IS_ACTIVE(CLOCK48MHZ_USE_PLLQ)) {
+            /* Enable PLLQ if PLL is used as 48MHz source clock */
+            RCC->PLLCFGR |= RCC_PLLCFGR_PLLQEN;
+        }
+
         RCC->CR |= (RCC_CR_PLLON);
         while (!(RCC->CR & RCC_CR_PLLRDY)) {}
+    }
 
-        /* now that the PLL is running, we use it as system clock */
+    /* Disable the current SYSCLK source (HSI), only if not used */
+    if (!IS_ACTIVE(CLOCK_ENABLE_HSI)) {
+        RCC->CFGR &= ~RCC_CFGR_SW;
+    }
+
+    /* Configure the system clock (SYSCLK) */
+    if (IS_ACTIVE(CONFIG_USE_CLOCK_HSE)) {
+        /* Select HSE as system clock */
+        RCC->CFGR |= RCC_CFGR_SW_HSE;
+        while ((RCC->CFGR & RCC_CFGR_SWS) != RCC_CFGR_SWS_HSE) {}
+    }
+    else if (IS_ACTIVE(CONFIG_USE_CLOCK_MSI)) {
+        /* Select MSI as system clock */
+        RCC->CFGR |= RCC_CFGR_SW_MSI;
+        while ((RCC->CFGR & RCC_CFGR_SWS) != RCC_CFGR_SWS_MSI) {}
+    }
+    else if (IS_ACTIVE(CONFIG_USE_CLOCK_PLL)) {
+#ifdef CPU_FAM_STM32L4
```

Comment:
```suggestion
#ifdef RCC_CFGR_HPRE_DIV2
```

Would that also make sense?

Reply:
The existence of the `RCC_CFGR_HPRE_DIV2` define doesn't mean that the clock warmup might be required, it's something more general that correspond to the HCLK prescaler and this define could exist for WB in future versions of the cmsis.
The clock warmup is something related to the L4 family (because WB is 64MHz max). So in this case, I think that keeping the family level check is more appropriate.